File: Interviews/LeetCode/Topic-Wise/Binary-Search/Medium/Minimum-Speed-to-Arrive-on-Time/1870.java
class Solution {
    public boolean f(int mid, int[] dist, double hour) {
        double totalTime = 0.0;
        for (int i = 0; i < dist.length - 1; i++) {
            totalTime += (dist[i] + mid - 1) / mid;
        }
        totalTime += (double) dist[dist.length - 1] / mid;
        return (totalTime <= hour);
    }

    public int minSpeedOnTime(int[] dist, double hour) {
        int answer = -1, low = 1, high = 10000000;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (f(mid, dist, hour)) {
                answer = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return answer;
    }
}

File: Interviews/LeetCode/Topic-Wise/Binary-Search/Medium/Minimum-Speed-to-Arrive-on-Time/1870.kt
class Solution {
    fun f(mid: Int, dist: IntArray, hour: Double): Boolean {
        var totalTime = 0.0
        for (i in 0 until dist.size - 1) {
            totalTime += ((dist[i] + mid - 1) / mid).toDouble()
        }
        totalTime += dist[dist.size - 1].toDouble() / mid
        return totalTime <= hour
    }

    fun minSpeedOnTime(dist: IntArray, hour: Double): Int {
        var answer = -1
        var low = 1
        var high = 10000000
        while (low <= high) {
            val mid = (low + high) / 2
            if (f(mid, dist, hour)) {
                answer = mid
                high = mid - 1
            } else {
                low = mid + 1
            }
        }
        return answer
    }
}